src/compiler/program.ts

```diff
@@ -1952,25 +1949,38 @@ namespace ts {
                 }
             }
 
-            function collectDynamicImportOrRequireCalls(node: Node): void {
-                if (isRequireCall(node, /*checkArgumentIsStringLiteralLike*/ true)) {
-                    imports = append(imports, node.arguments[0]);
-                }
-                // we have to check the argument list has length of 1. We will still have to process these even though we have parsing error.
-                else if (isImportCall(node) && node.arguments.length === 1 && isStringLiteralLike(node.arguments[0])) {
-                    imports = append(imports, node.arguments[0] as StringLiteralLike);
-                }
-                else if (isLiteralImportTypeNode(node)) {
-                    imports = append(imports, node.argument.literal);
-                }
-                collectDynamicImportOrRequireCallsForEachChild(node);
-                if (hasJSDocNodes(node)) {
-                    forEach(node.jsDoc, collectDynamicImportOrRequireCallsForEachChild);
+            function collectDynamicImportOrRequireCalls(file: SourceFile) {
+                const r = /import|require/g;

```

Comment:
you could tweak the regex to avoid false positives. though you need to adjust the loop a bit to save the result of `r.exec` to a variable.
```ts
/\b(?:import|require)\s*[(/]/g
```

The above will not match `require.resolve` for example as it requires an opening parenthesis or a slash (typically the start of a comment) after it.

Reply:
Since this PR is about performance, I'd rather we are careful to not regress performance where possible.
There is `require.resolve`, `import.meta`, in a project of mine there's a lot **require**sTypeInformation and in addition these are regular verbs that could occur in a lot of comments.
Each time you encounter something that contains one of these words, you need to recursively look up the node in the AST. That can potentially be very slow for large files with deep nesting.

Reply:
Well, the new collector is *m log(n)* where *m* is the number of `import|require` and *n* is the number of nodes. The old collector is *n log(n)*, and I think *m* will only approach *n* in pathological cases.

Also, the compiler is going to crash on very deep ASTs in the binder anyway. I tried to fix that by adding a flat OperatorListExpression to replace deep BinaryExpression trees, but gave up (for now) because of the amount of work needed to update all our BinaryExpression-based code.